segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+import io.pravega.segmentstore.contracts.AttributeUpdate;
+import io.pravega.segmentstore.contracts.AttributeUpdateType;
+import io.pravega.segmentstore.contracts.ReadResult;
+import io.pravega.segmentstore.contracts.ReadResultEntry;
+import io.pravega.segmentstore.server.ContainerEventProcessor;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+import io.pravega.segmentstore.server.SegmentStoreMetrics;
+import lombok.Cleanup;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * Implementation for {@link ContainerEventProcessor}. This class stores a map of {@link ContainerEventProcessor.EventProcessor}
+ * identified by name. Once this service is started, it will continuously perform processing iterations tailing the
+ * internal Segments of the registered {@link ContainerEventProcessor.EventProcessor}s (every ITERATION_SLEEP ms) and
+ * (safely) invoking their respective handler functions. This class durably stores the point at which the last successful
+ * events has been processed in the form of an attribute in the internal Segment (i.e., LAST_PROCESSED_EVENT_OFFSET).
+ * That is, before reading new events this attribute is read and after a successful processing of events this attribute
+ * is updated with the value of the next position to read. This class also truncates the internal Segments of the registered
+ * {@link ContainerEventProcessor.EventProcessor} based on that attribute.
+ */
+@Slf4j
+public final class ContainerEventProcessorImpl extends AbstractIdleService implements ContainerEventProcessor {
+
+    private static final UUID LAST_PROCESSED_EVENT_OFFSET = new UUID(0, 0);
+    private static final Duration ITERATION_SLEEP = Duration.ofMillis(100); // Execute 10 iterations per second at most.
+    private static final Duration OPERATION_TIMEOUT = Duration.ofMillis(20);
+
+    private final int containerId;
+    private final Map<String, EventProcessorImpl> eventProcessorMap = new ConcurrentHashMap<>();
+    private final Function<String, CompletableFuture<DirectSegmentAccess>> segmentSupplier;
+    private final AtomicBoolean closed;
+    private final ScheduledExecutorService executor;
+
+    //region Constructor
+
+    ContainerEventProcessorImpl(int containerId, @NonNull Function<String, CompletableFuture<DirectSegmentAccess>> segmentSupplier,
```

Comment:
Why is the supplier not `@NotNull`?

Reply:
Because having a `null` function from that we expect to invoke to get new `DirectSegmentAccess` objects is not going to be very useful.

Reply:
This pattern is convenient as it allows us to decouple the instantiation/implementation of internal Segments (e.g., mocks, real Segments) from the actual logic of the EventProcessor.